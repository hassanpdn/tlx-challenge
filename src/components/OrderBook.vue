<template>
      <base-card class="order-book full-width" title="Order Book">
            <base-table ref="sellItems" :items="sellItems" :headers="headers"></base-table>
            <base-card class="current-mark-price">Mark: {{numberSeperator(currentMarkPrice)}}</base-card>
            <base-table ref="buyItems" :items="buyItems" :headers="headers"></base-table>
      </base-card>
</template>

<script>
import BaseCard from "@/components/base/BaseCard.vue";
import BaseTable from "@/components/base/BaseTable.vue";
import { numberSeperator } from "@/helpers/helpers";
export default {
      name: "index",
      components: { BaseTable, BaseCard },
      data() {
            return {
                  // Web Socket state
                  connection: false,

                  // Table state
                  items: [],
                  headers: [
                        { title: `Price`, value: "price" },
                        { title: `Amount`, value: "amount" },
                  ],
                  price: 0,
                  amount: 0,
                  currentMarkPrice: 0,
            };
      },
      computed: {
            sellItems() {
                  return (
                        this.items?.existing?.filter(
                              (i) => i.side === "sell"
                        ) ||
                        this.items?.insert?.filter((i) => i.side === "sell")
                  );
            },
            buyItems() {
                  return (
                        this.items?.existing?.filter((i) => i.side === "buy") ||
                        this.items?.insert?.filter((i) => i.side === "buy")
                  );
            },
      },
      methods: {
            numberSeperator,
            averageCalculation() {
                  const max = Math.max(...this.sellItems.map((i) => i.price));
                  const min = Math.min(...this.buyItems.map((i) => i.price));

                  this.currentMarkPrice = (min + max) / 2;
            },
      },
      created() {
            console.log("Starting connection to WebSocket Server");
            this.connection = new WebSocket("ws://localhost:8080");
            this.connection.onopen = function (event) {
                  console.log(event);
                  console.log("Successfully connected to WebSocket");
            };

            this.connection.onmessage = ({ data }) => {
                  const message = JSON.parse(data);
                  this.items = message;

                  this.averageCalculation();
            };
      },
      async mounted() {
            this.$nextTick(() => {
                  this.$refs.sellItems.scrollToBottom();
            });
      },
      beforeDestroy() {
            this.connection.close();
      },
};
</script>

<style>
      .order-book {
            margin: 40px;
      }

      .current-mark-price {
            margin: 25px 0;
      }
</style>