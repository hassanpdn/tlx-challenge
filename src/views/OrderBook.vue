<template>
      <base-card class="order-book full-width" title="Order Book">
            <template slot="cardHeader">
                  <div class="card-fix-header">
                        <span>Order Book</span>
                        <base-select v-model="groupByNumber" @select="select" :list="prices"></base-select>
                  </div>
            </template>
            <base-table @rowClick="onItemClick" ref="sellOrders" :items="sellOrdersGroupBy" :headers="headers"></base-table>
            <base-card class="current-mark-price">Mark: {{priceSeperator(currentMarkPrice) || '- - -'}}</base-card>
            <base-table @rowClick="onItemClick" ref="buyOrders" :items="buyOrdersGroupBy" :headers="headers"></base-table>
      </base-card>
</template>

<script>

// Components
import BaseCard from "@/components/base/BaseCard.vue";
import BaseTable from "@/components/base/BaseTable.vue";
import BaseSelect from "@/components/base/BaseSelect.vue";

// Helpers
import { priceSeperator } from "@/helpers/helpers";
import { webSocket } from "@/services/ws";

export default {
      name: "index",
      components: { BaseTable, BaseCard, BaseSelect },
      data() {
            return {
                  // Web Socket state
                  connection: false,

                  // Table state
                  items: [],
                  headers: [
                        { title: `Price`, value: "price" },
                        { title: `Amount`, value: "amount" },
                  ],
                  currentMarkPrice: 0,

                  // Select state
                  prices: [10, 20, 30, 40],
                  groupByNumber: null
            };
      },
      computed: {
            buyOrdersGroupBy() {
                  const buyOrders = this.items.filter((order) => order.side === "buy");
                  const sortedBuyOrders = this.sortOrders(buyOrders);
                  if (Number(this.groupByNumber)) {
                        return this.groupBy(sortedBuyOrders);
                  }
                  return sortedBuyOrders;
            },
            sellOrdersGroupBy() {
                  const sellOrders = this.items.filter((order) => order.side === "sell");
                  const sortedSellOrders = this.sortOrders(sellOrders);
                  if (Number(this.groupByNumber)) {
                        return this.groupBy(sortedSellOrders);
                  }
                  return sortedSellOrders;
            }
      },
      methods: {
            priceSeperator,
            averageCalculation() {
                  const min_ask = Math.min(
                        ...this.sellOrdersGroupBy.map((o) => o.price)
                  );
                  const max_bid = Math.max(
                        ...this.buyOrdersGroupBy.map((o) => o.price)
                  );

                  if (isFinite(min_ask) && isFinite(max_bid)) {
                        this.currentMarkPrice = document.title = ((min_ask + max_bid) * 0.5).toFixed(2);
                  } else if (isFinite(min_ask)) {
                        this.currentMarkPrice = document.title = min_ask.toFixed(2);
                  } else if (isFinite(max_bid)) {
                        this.currentMarkPrice = document.title = max_bid.toFixed(2);
                  } else {
                        this.currentMarkPrice = document.title = '';
                  }

            },
            sortOrders(arr){
                  return arr.sort((a, b) => parseFloat(a.price) - parseFloat(b.price));
            },
            onItemClick(item){
                  this.$root.$emit('setOrder', item)
            },
            select(val){
                  this.groupByNumber = val;
            },
            groupBy(array) {
                  const buyOrdersLength = array.length;
                  let finalResults = [];
                  let i = buyOrdersLength - 1;
                  const priceInteger = Math.trunc(array[i].price);
                  let maxPrice = priceInteger - (priceInteger % Number(this.groupByNumber)) + Number(this.groupByNumber);
                  while (i >= 0) {
                        let amountSum = 0;
                        let j = i;
                        while (j >= 0 && array[j].price <= maxPrice) {
                              amountSum += array[j].amount;
                              j--;
                        }
                        i = j;
                        finalResults.unshift({
                              price: maxPrice,
                              amount: amountSum,
                        });
                        maxPrice += Number(this.groupByNumber);
                  }
                  return finalResults;
            },
            handleOrdersInsertion(arr){
                  if(arr.existing) {
                        this.items = arr.existing;
                  } 
                  if(arr.delete) {
                        this.items = this.items.filter(order => !arr.delete.includes(order.id))
                  }
                  if(arr.insert) {
                        this.items.push(...arr.insert)
                  }
            },
            startSocket(){
                  console.log("Starting connection to WebSocket Server");
                  this.connection = webSocket;
                  this.connection.onopen = function (event) {
                        console.log(event);
                        console.log("Successfully connected to WebSocket");
                  };

                  this.connection.onmessage = ({ data }) => {
                        const message = JSON.parse(data);
                        this.handleOrdersInsertion(message);
                        this.averageCalculation();
                  };
            }
      },
      created() {
            this.startSocket();

      },
      beforeDestroy() {
            this.connection.close();
      },
};
</script>

<style>

.current-mark-price {
      margin: 25px 0;
}
</style>