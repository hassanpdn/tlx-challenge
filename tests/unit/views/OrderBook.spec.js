
import { mount } from '@vue/test-utils';
import OrderBook from '@/views/OrderBook';

describe('OrderBook', () => {
      test("renders component", () => {
            const wrapper = mount(OrderBook);
            expect(wrapper.element).toMatchSnapshot()
      })

      test('averageCalculation', () => {
            const data = {'insert':[{price: 20, amount: 26.181920, side: 'buy'}, {price: 80, amount: 26.181920, side: 'sell'}, {price: 15, amount: 26.181920, side: 'sell'}, {price: 30, amount: 26.181920, side: 'buy'}]};
            const wrapper = mount(OrderBook);
            wrapper.setData({items: data})
            wrapper.vm.averageCalculation();
            expect(wrapper.vm.currentMarkPrice).toEqual(22.5);
      })
});
