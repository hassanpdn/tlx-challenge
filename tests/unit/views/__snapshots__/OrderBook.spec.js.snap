// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderBook renders component 1`] = `
<div
  class="card order-book full-width"
>
  <div
    class="card-header"
  >
    <div
      class="card-fix-header"
    >
      <span>
        Order Book
      </span>
       
      <div>
        <label
          for="base-select"
        >
          
        </label>
         
        <select
          class="select"
          id="base-select"
        >
          <option
            selected="selected"
            value="null"
          >
            Select a group size
          </option>
           
          <option
            class="select-dropdown__list-item"
            value="10"
          >
            10
          </option>
          <option
            class="select-dropdown__list-item"
            value="20"
          >
            20
          </option>
          <option
            class="select-dropdown__list-item"
            value="30"
          >
            30
          </option>
          <option
            class="select-dropdown__list-item"
            value="40"
          >
            40
          </option>
        </select>
      </div>
    </div>
  </div>
   
  <div
    class="card-body"
  >
     
    <div
      class="root"
      style="height: 200px; overflow: auto;"
    >
      <div
        class="viewport"
        style="overflow: hidden; height: 150px; position: relative;"
      >
        <div
          class="spacer"
          style="transform: translateY(0px);"
        >
          <table
            class="table"
          >
            <thead>
              <tr>
                <th
                  class="row-number"
                >
                  Price
                </th>
                <th
                  class="row-number"
                >
                  Amount
                </th>
              </tr>
            </thead>
             
            <tbody />
          </table>
        </div>
      </div>
    </div>
     
    <div
      class="card current-mark-price"
    >
      <!---->
       
      <div
        class="card-body"
      >
        Mark: 0
      </div>
    </div>
     
    <div
      class="root"
      style="height: 200px; overflow: auto;"
    >
      <div
        class="viewport"
        style="overflow: hidden; height: 150px; position: relative;"
      >
        <div
          class="spacer"
          style="transform: translateY(0px);"
        >
          <table
            class="table"
          >
            <thead>
              <tr>
                <th
                  class="row-number"
                >
                  Price
                </th>
                <th
                  class="row-number"
                >
                  Amount
                </th>
              </tr>
            </thead>
             
            <tbody />
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
